frontend/src/Frontend/UI/Widgets.hs

```diff
@@ -176,59 +177,102 @@ uiRealInputElement cfg = do
     inputElement $ cfg & initialAttributes %~
         (<> ("type" =: "number")) . addInputElementCls . addNoAutofillAttrs
 
+uiCorrectingInputElement
+  :: forall t m a er explanation. DomBuilder t m
+  => MonadFix m
+  => (Text -> Maybe a)
+  -> (a -> Maybe (a, explanation))
+  -> (a -> Text)
+  -> InputElementConfig er t (DomBuilderSpace m)
+  -> m (InputElement er (DomBuilderSpace m) t, Dynamic t (Maybe a), Event t (a, Maybe explanation))
+uiCorrectingInputElement parse sanitize render cfg = mdo
+  ie <- inputElement $ cfg & inputElementConfig_setValue %~ (\e -> leftmost [attemptCorrection e, forceCorrection inp])
+  let
+    inp = _inputElement_input ie
+    val = value ie
+
+    sanitization :: Functor f => f Text -> f (Maybe (a, Maybe explanation))
+    sanitization = fmap $ \i -> ffor (parse i) $ \p ->
+      case sanitize p of
+        Nothing -> (p, Nothing)
+        Just (a, x) -> (a, Just x)
+
+    intercept :: ((a, Maybe explanation) -> Bool) -> (Event t Text -> Event t Text)
+    intercept p = fmap render . fmap fst . ffilter p . fmapMaybe id . sanitization
+
+    attemptCorrection :: Event t Text -> Event t Text
+    attemptCorrection = intercept (const True)
+
+    forceCorrection :: Event t Text -> Event t Text
+    forceCorrection = intercept (isJust . snd)
+
+    -- type inference goes crazy if we inline this - ghc bug?
+    inp' = fmapMaybe id $ sanitization inp
+
+  pure (ie, (fmap . fmap) fst $ sanitization val, inp')
+
 -- | Decimal input to the given precision. Returns the element, the value, and
 -- the user input events
-uiRealWithPrecisionInputElement
+uiNonnegativeRealWithPrecisionInputElement
   :: forall t m er a. (DomBuilder t m, MonadFix m, MonadHold t m)
   => Word8
   -> (Decimal -> a)
   -> InputElementConfig er t (DomBuilderSpace m)
-  -> m (InputElement er (DomBuilderSpace m) t, (Dynamic t (Maybe a), Event t a))
-uiRealWithPrecisionInputElement prec fromDecimal cfg = do
+  -> m (InputElement er (DomBuilderSpace m) t, Dynamic t (Maybe a), Event t a)
+uiNonnegativeRealWithPrecisionInputElement prec fromDecimal cfg = do
```

Comment:
I hate this name

Reply:
It gets slightly better if you change nonnegative -> positive, but yeah

Reply:
It accepts 0 though.